File: frontend/src/lib/types/analytics.ts
export type AnalyticsScope = "user" | "organization";

export type TimeSeriesPoint = { bucket: string; value: number };
export type CountSeriesPoint = { bucket: string; count: number };
export type LabelCount = { label: string; count: number };

export type AnalyticsSummary = {
  violation_events: number;
  blocked_executions: number;
  policy_violations: number;
  prompt_scans_total: number;
  prompt_blocked: number;
  output_scans_total: number;
  output_blocked: number;
  avg_prompt_risk: number | null;
  avg_output_risk: number | null;
  unread_notifications: number;
  scope: AnalyticsScope;
};

export type RealtimeViolation = {
  id: string;
  event_type: string;
  severity: string;
  occurred_at: string;
  user_id: string | null;
  resource_type: string | null;
  resource_id: string | null;
  payload?: Record<string, unknown> | null;
};

export type PromptMonitoringStats = {
  total_scans: number;
  blocked: number;
  warned: number;
  allowed: number;
  decision_breakdown: LabelCount[];
  avg_risk_score: number | null;
};

export type OutputLeakageStats = {
  total_scans: number;
  blocked: number;
  warned: number;
  leakage_breakdown: LabelCount[];
  avg_risk_score: number | null;
};

export type BlockedExecutionsStats = {
  total_blocked: number;
  status_breakdown: LabelCount[];
  trend: CountSeriesPoint[];
};

export type HighRiskUser = {
  user_id: string;
  email: string;
  full_name: string | null;
  scan_count: number;
  avg_risk_score: number;
  blocked_prompts: number;
};

export type HighRiskModel = {
  model_id: string;
  name: string;
  provider: string | null;
  execution_count: number;
  blocked_count: number;
};

export type AnalyticsDashboard = {
  summary: AnalyticsSummary;
  execution_trend: CountSeriesPoint[];
  risk_trend: TimeSeriesPoint[];
  violation_trend: CountSeriesPoint[];
  policy_violation_trend: CountSeriesPoint[];
  prompt_stats: PromptMonitoringStats;
  output_stats: OutputLeakageStats;
  blocked_executions: BlockedExecutionsStats;
  realtime_violations: RealtimeViolation[];
  high_risk_users: HighRiskUser[];
  high_risk_models: HighRiskModel[];
  guard_actions: LabelCount[];
};

export type AnalyticsFilters = {
  days: number;
  severity: string;
  granularity: "hour" | "day" | "week";
};

File: backend/app/schemas/analytics.py
"""Analytics API schemas (Sprint 3 Step 7)."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class TimeSeriesPoint(BaseModel):
    bucket: datetime
    value: float


class CountSeriesPoint(BaseModel):
    bucket: datetime
    count: int


class LabelCount(BaseModel):
    label: str
    count: int


class AnalyticsSummaryResponse(BaseModel):
    violation_events: int = 0
    blocked_executions: int = 0
    policy_violations: int = 0
    prompt_scans_total: int = 0
    prompt_blocked: int = 0
    output_scans_total: int = 0
    output_blocked: int = 0
    avg_prompt_risk: float | None = None
    avg_output_risk: float | None = None
    unread_notifications: int = 0
    scope: str = Field(description="'user' or 'organization'")


class RealtimeViolationItem(BaseModel):
    id: UUID
    event_type: str
    severity: str
    occurred_at: datetime
    user_id: UUID | None
    resource_type: str | None
    resource_id: UUID | None
    payload: dict | None = None


class RealtimeViolationsResponse(BaseModel):
    items: list[RealtimeViolationItem]
    total: int


class TrendSeriesResponse(BaseModel):
    metric: str
    granularity: str
    points: list[CountSeriesPoint]


class RiskTrendResponse(BaseModel):
    granularity: str
    points: list[TimeSeriesPoint]


class PromptMonitoringStatsResponse(BaseModel):
    total_scans: int
    blocked: int
    warned: int
    allowed: int
    decision_breakdown: list[LabelCount]
    avg_risk_score: float | None = None


class OutputLeakageStatsResponse(BaseModel):
    total_scans: int
    blocked: int
    warned: int
    leakage_breakdown: list[LabelCount]
    avg_risk_score: float | None = None


class BlockedExecutionsResponse(BaseModel):
    total_blocked: int
    status_breakdown: list[LabelCount]
    trend: list[CountSeriesPoint]


class HighRiskUserItem(BaseModel):
    user_id: UUID
    email: str
    full_name: str | None
    scan_count: int
    avg_risk_score: float
    blocked_prompts: int


class HighRiskUsersResponse(BaseModel):
    items: list[HighRiskUserItem]


class HighRiskModelItem(BaseModel):
    model_id: UUID
    name: str
    provider: str | None
    execution_count: int
    blocked_count: int


class HighRiskModelsResponse(BaseModel):
    items: list[HighRiskModelItem]


class PolicyViolationTrendResponse(BaseModel):
    total: int
    points: list[CountSeriesPoint]


class AnalyticsDashboardResponse(BaseModel):
    """Combined payload for the monitoring dashboard page."""

    summary: AnalyticsSummaryResponse
    execution_trend: list[CountSeriesPoint]
    risk_trend: list[TimeSeriesPoint]
    violation_trend: list[CountSeriesPoint]
    policy_violation_trend: list[CountSeriesPoint]
    prompt_stats: PromptMonitoringStatsResponse
    output_stats: OutputLeakageStatsResponse
    blocked_executions: BlockedExecutionsResponse
    realtime_violations: list[RealtimeViolationItem]
    high_risk_users: list[HighRiskUserItem]
    high_risk_models: list[HighRiskModelItem]
    guard_actions: list[LabelCount]
